### ISSUE TITLE
Difference of interpretation between Implicit Class Target and sh:class constraint

### ISSUE THREAD

**Issue by tfrancart** [TopQuadrant/shacl#76]
Given the following test data

```
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .

<http://example.com/ontology#PaintingActivity>
	rdfs:subClassOf <http://example.com/ontology#Activity> ;
.

<http://example.com/PaintingActivity/activity_1>
        a       <http://example.com/ontology#PaintingActivity> ;
        <http://example.com/ontology#activityHasStatus>
                <http://example.com/ontology/done> ;
.
```

And the foloowing test shapes :

```
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix sh:    <http://www.w3.org/ns/shacl#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix skos:  <http://www.w3.org/2004/02/skos/core#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

<http://example.com/ontology#activityHasStatus-shape>
        a                    sh:NodeShape ;
        sh:class             <http://example.com/ontology#Activity> ;
        sh:targetSubjectsOf  <http://example.com/ontology#activityHasStatus> .

<http://example.com/ontology#Activity>
        a sh:NodeShape, rdfs:Class ;
        sh:property [
                sh:class  <http://example.com/ontology#ActivityStatus> ;
        ]
.
```

(Note that there is NodeShape with an implicit target as it is using "a sh:NodeShape, rdfs:Class ;", and a shape that is using the sh:class constraint on the same class "Activity")

A validation produces the following 2 validation results :

```
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix sh:    <http://www.w3.org/ns/shacl#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix skos:  <http://www.w3.org/2004/02/skos/core#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a            sh:ValidationReport ;
  sh:conforms  false ;
  sh:result    [ a                             sh:ValidationResult ;
                 sh:focusNode                  <http://example.com/PaintingActivity/activity_1> ;
                 sh:resultMessage              "Value must be an instance of <http://example.com/ontology#ActivityStatus>" ;
                 sh:resultSeverity             sh:Violation ;
                 sh:sourceConstraintComponent  sh:ClassConstraintComponent ;
                 sh:sourceShape                []  ;
                 sh:value                      <http://example.com/PaintingActivity/activity_1>
               ] ;
  sh:result    [ a                             sh:ValidationResult ;
                 sh:focusNode                  <http://example.com/PaintingActivity/activity_1> ;
                 sh:resultMessage              "Value must be an instance of <http://example.com/ontology#Activity>" ;
                 sh:resultSeverity             sh:Violation ;
                 sh:sourceConstraintComponent  sh:ClassConstraintComponent ;
                 sh:sourceShape                <http://example.com/ontology#activityHasStatus-shape> ;
                 sh:value                      <http://example.com/PaintingActivity/activity_1>
               ]
] .
```

Note that :

- The shape that uses implicit targeting is applied because the property shape attached to it produces a validation result, hence the path `:activity_1 rdf:type PaintingActivity + PaintingActovity rdfs:subclassOf Activity` is properly interpreted here
- However, the `sh:class` constraint on the other shape generates a validation result, while I would expect it does not, because `activity_1` is an indirect instance of Activity, as proven by the other validation result : `:activity_1 rdf:type PaintingActivity + PaintingActovity rdfs:subclassOf Activity`

What can be the reason for this ?

I am using release 1.3.0, and validating in a custom code of mine.

**Comment by tfrancart** [comment]
Validating using `shaclvalidate.sh` produces the same output :

```
thomas@thomas-ThinkPad-T460s:~/sparna/00-Clients/Chut/04-SHACL$ shaclvalidate.sh -datafile 02-Data/test.ttl -shapesfile 03-Shapes/test-shapes.ttl 
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a       <http://www.w3.org/ns/shacl#ValidationReport> ;
  <http://www.w3.org/ns/shacl#conforms>
          false ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/PaintingActivity/activity_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value must be an instance of <http://example.com/ontology#ActivityStatus>" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#ClassConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    []  ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/PaintingActivity/activity_1>
          ] ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/PaintingActivity/activity_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value must be an instance of <http://example.com/ontology#Activity>" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#ClassConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    <http://example.com/ontology#activityHasStatus-shape> ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/PaintingActivity/activity_1>
          ]
] .
```

**Comment by irenetq** [comment]
Looks like a malformed property shape - no sh:path

**Comment by tfrancart** [comment]
OK but this was not my point. Even after fixing this, I still a difference between the interpreation of `.... a sh:NodeShape, rdfs:Class`and the interpretation of `... a sh:NodeShape; sh:class X`.

Here is the fixed shape file :

```
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix sh:    <http://www.w3.org/ns/shacl#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix skos:  <http://www.w3.org/2004/02/skos/core#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

<http://example.com/ontology#activityHasStatus-shape>
        a                    sh:NodeShape ;
        sh:class             <http://example.com/ontology#Activity> ;
        sh:targetSubjectsOf  <http://example.com/ontology#activityHasStatus> .

<http://example.com/ontology#Activity>
        a sh:NodeShape, rdfs:Class ;
        sh:property [
                sh:property <http://example.com/ontology#activityHasStatus> ;
                sh:class  <http://example.com/ontology#ActivityStatus> ;
        ]
.
```

And the output of shaclvalidate.sh :

```
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a       <http://www.w3.org/ns/shacl#ValidationReport> ;
  <http://www.w3.org/ns/shacl#conforms>
          false ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/PaintingActivity/activity_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value must be an instance of <http://example.com/ontology#ActivityStatus>" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#ClassConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    []  ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/PaintingActivity/activity_1>
          ] ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/PaintingActivity/activity_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value must be an instance of <http://example.com/ontology#Activity>" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#ClassConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    <http://example.com/ontology#activityHasStatus-shape> ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/PaintingActivity/activity_1>
          ]
] .
```

To rephrase my question : why is the sh:class constraint on a NodeShape triggering a ValidationResult even though the node in the test data is indirectly an instance of that Class, as proven by the fact that the same Class is properly used as implicit target by the other shape ?

**Comment by tfrancart** [comment]
Still simplifying using sh:nodeKind as a proof the implicit target class is taken into account, instead of a property shape, to avoid confusion :

```
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix sh:    <http://www.w3.org/ns/shacl#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix skos:  <http://www.w3.org/2004/02/skos/core#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

<http://example.com/ontology#activityHasStatus-shape>
        a                    sh:NodeShape ;
        sh:class             <http://example.com/ontology#Activity> ;
        sh:targetSubjectsOf  <http://example.com/ontology#activityHasStatus> .

<http://example.com/ontology#Activity>
        a sh:NodeShape, rdfs:Class ;
        sh:nodeKind sh:BlankNode ;
.
```

and the output :

```
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a       <http://www.w3.org/ns/shacl#ValidationReport> ;
  <http://www.w3.org/ns/shacl#conforms>
          false ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/PaintingActivity/activity_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value does not have node kind BlankNode" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#NodeKindConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    <http://example.com/ontology#Activity> ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/PaintingActivity/activity_1>
          ] ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/PaintingActivity/activity_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value must be an instance of <http://example.com/ontology#Activity>" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#ClassConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    <http://example.com/ontology#activityHasStatus-shape> ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/PaintingActivity/activity_1>
          ]
] .

```

**Comment by HolgerKnublauch** [comment]
I believe this was caused by a regression in 1.3.0. I had introduced a cache to make sh:class faster but that cache operated on the class hierarchy from the shapes graph, not the data graph.

Would you be able to try the latest 1.3.1 SNAPSHOT (i.e. build from GitHub) and let me know if that has fixed it?

**Comment by tfrancart** [comment]
Indeed, using 1.3.1-SNAPSHOT produces the expected result, that is only one violation :

```
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a       <http://www.w3.org/ns/shacl#ValidationReport> ;
  <http://www.w3.org/ns/shacl#conforms>
          false ;
  <http://www.w3.org/ns/shacl#result>
          [ a       <http://www.w3.org/ns/shacl#ValidationResult> ;
            <http://www.w3.org/ns/shacl#focusNode>
                    <http://example.com/instance_1> ;
            <http://www.w3.org/ns/shacl#resultMessage>
                    "Value does not have node kind BlankNode" ;
            <http://www.w3.org/ns/shacl#resultSeverity>
                    <http://www.w3.org/ns/shacl#Violation> ;
            <http://www.w3.org/ns/shacl#sourceConstraintComponent>
                    <http://www.w3.org/ns/shacl#NodeKindConstraintComponent> ;
            <http://www.w3.org/ns/shacl#sourceShape>
                    <http://example.com/ontology#SuperClass> ;
            <http://www.w3.org/ns/shacl#value>
                    <http://example.com/instance_1>
          ]
] .
```

Can you confirm it is OK in general to have the class hierarchy as part of the data graph and not the shapes graph ?

**Comment by HolgerKnublauch** [comment]
sh:class looks for subClassOf triples in the data graph, see its definition: For each value node that is either a literal, or a non-literal that is not a SHACL instance of $class in the data graph, there is a validation result with the value node as sh:value.

The same applies to sh:targetClass: If s is a shape in a shapes graph SG and s has value c for sh:targetClass in SG then the set of SHACL instances of c in a data graph DG is a target from DG for s in SG.

And correspondingly to implicit class targets.

If you find the ticket addressed please close it. Thanks for your bug report.

**Comment by tfrancart** [comment]
Of course the ticket is adressed. Thanks

**Comment by atextor** [comment]
I was just hit by this regression when updating to 1.3.0. Good to know that this is the cause.
@HolgerKnublauch Is there a schedule for a 1.3.1 release that comes with the fix?

**Comment by HolgerKnublauch** [comment]
Yes given that this was a real bug in the semantics I want to do an official release on Monday or so.

**Comment by HolgerKnublauch** [comment]
I have released 1.3.1. I hope it works for everyone.
